### DISCUSSION TITLE
Is USB_to_USB converter by Hasu supported?

### DISCUSSION THREAD

**Discussion by bogorad** [manna-harbour/miryoku#303] [Q&A] [answered]
It's definitely supported by QMK (`converters/usb_usb/hasu`). What steps do I need to take to compile Miryoku for it?

**Comment by manna-harbour** [comment, marked as answer]
First, for row stagger, there are 3 options. See https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/users/manna-harbour_miryoku#60_ansi for details. So you need to decide on one or more of those options.

https://github.com/qmk/qmk_firmware/tree/master/keyboards/converter/usb_usb/keymaps/manna-harbour_miryoku/config.h needs to be created, based on https://github.com/qmk/qmk_firmware/blob/master/keyboards/converter/usb_usb/keymaps/default/keymap.c and https://github.com/manna-harbour/miryoku_qmk/blob/miryoku/layouts/community/60_ansi/manna-harbour_miryoku/config.h. See https://github.com/manna-harbour/miryoku/discussions/264.

Also see https://github.com/manna-harbour/miryoku/discussions/287.

As an alternative to Miryoku QMK on a usb-usb converter, see https://github.com/manna-harbour/miryoku_kmonad and https://github.com/manna-harbour/xmk.

**Reply by bogorad** [reply]
Many thanks for a quick response and the pointers! I think I got the idea.

I started with `miryoku_kmonad`, then re-created the config for [Kanata](https://github.com/jtroo/kanata) (it has chords and proper mouse control) and have been using it exclusively with my Kinesis Advantage (KB500), but still prefer a hardware solution, hence the converter.

I know about the [kint](https://github.com/kinx-project/kint), waiting for parts.

Also, I have a Piantor Pro now and would like a single keymap config for both of my keyboards.

**Reply by manna-harbour** [reply]
Great! Let me know if you get stuck.

If it's for use with a KA, you could add a layout for it to https://github.com/qmk/qmk_firmware/blob/master/keyboards/converter/usb_usb/info.json, then you'd use that layout for the Miryoku mapping, which could look something like the ADV360 mapping from Miryoku ZMK: https://github.com/manna-harbour/miryoku_zmk/blob/master/miryoku/mapping/76/adv360.h.

Nice! I'd like to add a Kanata option to Miryoku KMonad at some point.

**Reply by bogorad** [reply]
Now I'm stuck :)

I added the definition to `qmk_firmware/keyboards/converter/usb_usb/info.json`, then created the matrix in `qmk_firmware/keyboards/converter/usb_usb/usb_usb.h`, then created `qmk_firmware/keyboards/converter/usb_usb/keymaps/manna-harbour_miryoku/config.h`.

But when I run:

```
qmk compile -kb converter/usb_usb/hasu -km manna-harbour_miryoku -e MIRYOKU_ALPHAS=COLEMAKDH  -e MIRYOKU_EXTRA=QWERTY  -e MIRYOKU_TAP=QWERTY   -e MIRYOKU_CLIPBOARD=WIN
```

I get [these errors](https://zerobin.org/?8f142da2bbd1e3d4#8xroGZXUzduR1efGezzF5JRm8coZjrYjU6e2ZpNHTBFn).

Namely, "macro "LAYOUT_miryoku" passed 40 arguments but takes just 36" while I'm definitely passing 36.

Could you take a look, where did I screw up? All files are here:

https://github.com/bogorad/usb_usb_kinesis_miryoku

**Reply by manna-harbour** [reply]
If you've split the halves in the layout in info.json, then you'll need to also split in https://github.com/bogorad/usb_usb_kinesis_miryoku/blob/d59f30f83c7b035b1fceb780760b3e8ee1384992/keymaps/manna-harbour_miryoku/config.h#L21C1-L39C2 as you have done, but you shouldn't split https://github.com/bogorad/usb_usb_kinesis_miryoku/blob/d59f30f83c7b035b1fceb780760b3e8ee1384992/keymaps/manna-harbour_miryoku/config.h#L11C1-L20C1. In particular, that section needs to have the same form as everywhere else e.g. https://github.com/manna-harbour/miryoku_zmk/blob/dc6f0e921c515e9ba39c6ec79bb669810102799b/miryoku/mapping/76/adv360.h#L9C1-L12C101, as it is called using the layer data (which has padding on the bottom row, hence the 40) e.g. https://github.com/manna-harbour/miryoku_zmk/blob/master/miryoku/miryoku_babel/miryoku_layer_alternatives.h. You can change the names of the arguments if desired, but not the number and order.

**Reply by manna-harbour** [reply]
BTW, I'd expect you don't need to define the layout in both usb_usb.h and info.json.

**Reply by bogorad** [reply]
Thank you for explaining!

Indeed, when reading the QMK doc, I missed the "data-driven" part in the [info.json refence](https://github.com/qmk/qmk_firmware/blob/master/docs/reference_info_json.md). QMK was complaining about the missing `matrix` definition, and I just defined it in the `usb_usb.h`. This was needless, so I went back and [re-did my `info.json` so now it includes `matrix` directives for each key](https://github.com/bogorad/usb_usb_kinesis_miryoku/blob/main/info.json) - `LAYOUT_kinesis_advantage`.

Also, thanks for pointing out the need for padding. Now it compiles without errors!

However, [now it refuses to link](https://zerobin.org/?be045ee878b43bc3#B76yLTYFwYG1Yp5T3Zc312y9ewUfYbEhswWS5Dqwtum). I assume atmega32u4 doesn't have enough flash? Am I out of luck?

**Reply by manna-harbour** [reply]
See https://github.com/qmk/qmk_firmware/blob/master/docs/squeezing_avr.md.

In this case, the custom matrix is extremely large, so reducing the size of the matrix, or reducing the number of layers should help. You could first try adding or removing a layer from the default keymap to see if it's significant.

For Miryoku you can leave out unused layers. See https://github.com/manna-harbour/miryoku/discussions/85.

For your layout, half the matrix rows are unused, so you could modify the custom matrix code to skip those.

Otherwise, Miryoku QMK has a few QMK features enabled that aren't critical and could be disabled, e.g. by editing https://github.com/manna-harbour/miryoku_qmk/blob/miryoku/users/manna-harbour_miryoku/rules.mk and https://github.com/manna-harbour/miryoku_qmk/blob/miryoku/users/manna-harbour_miryoku/manna-harbour_miryoku.c.

**Reply by manna-harbour** [reply]
Or use the approach from https://github.com/qmk/qmk_firmware/tree/master/keyboards/converter/xmk, and have the matrix represent only the keycodes used for input.

**Reply by bogorad** [reply]
Thanks! I decided to put this on hold until I better understand QMK. I'll be back with more questions :)

**Reply by bogorad** [reply]
Well, I felt comfortable enough to start rebuilding it from the ground up, and realized that the moment I enable both tap-dance and combos, I run out of space. Oh well. I'll just have to wait for the pcbs/chips for Kint to arrive.

Thanks again for your assistance!

**Reply by manna-harbour** [reply]
Changing the MCU will be the best option in general.

Tap dance is only used for the double tap guard, so that can be removed.

If you want to try the `converter/xmk` approach, you'd add a mapping array to map from input keycode to keymap key position, equivalent to the layout macro in the usb converter config.
